### DISCUSSION TITLE
testing reusable workflows locally - replace remote repo with local path?

### DISCUSSION THREAD

**Discussion by mverkerk-godaddy** [nektos/act#1924] [Q&A]
Hi All - I'm having an issue while testing reusable workflows locally. Here's an example:

```
jobs:
  reusable:
    uses: org/repo/.github/workflows/reusable.yml@main
```
When I run this job, act will clone `https://github.com/org/repo` to local cache and throw an error: `Error: stat /home/me/.cache/act/org-repo@main/.github/workflows/reusable.yml: no such file or directory`

This _is_ the expected behavior but instead if this, I want act to use the local version of the file because that is what I am testing or, in this case, the file has not been released yet ...

Of course, I could change to `./.github/workflows/reusable.yml` while testing but that can easily lead to mistakes (i.e. forgot to change back the path) or I may have many of these and would have to change all the paths back & forth

Would be great if there was a parameter like `--remote-repo <org>/<repo>` which act would interpret as `./`

Is there a way to do this already?

**Comment by ChristopherHX** [comment]
No such feature exists in act, no doubt this is useful.
I would propose the following once https://github.com/nektos/act/pull/1913 and it's related changes get merged (Full restructure of the whole Action Cacheing code and removal of hardcoded paths on the host)
```
--local-repository <local-repository>                                    Redirect dependent repositories to the local filesystem. E.g `--local-repository org/name@ref=/path/to/repository`
```

_I have already implemented this in my actions/runner fork, but act may gain it later_

**Reply by verkurkie** [reply]
That would be awesome! i will keep and eye on https://github.com/nektos/act/pull/1913 and propose the new parameter after that has been merged! Thx for the response 🤞

**Reply by dekimsey** [reply]
@ChristopherHX just making sure I'm following the right PRs for status. It looks like #1918 merged in August 2023 (yay!), does #1954 reflect this feature request?

**Reply by ChristopherHX** [reply]
@dekimsey 
> does https://github.com/nektos/act/pull/1954 reflect this feature request?

No the pr is about remote actions fail to update with the current logic, but It's the only direct required dependency including enabling my new layered architecture of remote actions.

If you want to try it out, please have a look at this branch: <https://github.com/nektos/act/tree/local-repository-action-cache> (aka https://github.com/ChristopherHX/act/pull/34, due to how GitHub works I couldn't open it in the parent repo, because it's dependency lives in my fork).

**Reply by ChristopherHX** [reply]
I have opened PR https://github.com/nektos/act/pull/2155 to raise visibility, but I don't expect that this get's merged in a 6months timeframe

pr #2155 does implement this feature, at least it worked 6months ago.

**Reply by brianjmurrell** [reply]
When I specify:

```
 --local-repository my-company/actions-lib/.github/workflows/build.yml@foo/initial=/home/brian/bar/actions-lib/
```

Is what is used the "working copy" (i.e. the directory contents including uncommitted changes) of what's in `/home/brian/bar/actions-lib/` or is it what is on the `foo/initial` branch in `/home/brian/bar/actions-lib/`?

To clarify, do I need to actually commit everything in  `/home/brian/bar/actions-lib/` in order for _act_ to use what is in that directory as the local repository?  Ideally I'd like to avoid committing in `/home/brian/bar/actions-lib/` until I have tested it with act rather than having a bazillion commits there for every different change I might want to test while homing on the final solution.

If it is the case of having to commit, perhaps could there be some kind of pseudo tag (i.e. `--local-repository my-company/actions-lib/.github/workflows/build.yml@foo/initial=/home/brian/bar/actions-lib@working-copy` or something) that could tell _act_ to use the contents of the files in the directory rather than what's been committed to a branch?

**Reply by ChristopherHX** [reply]
> do I need to actually commit everything in /home/brian/bar/actions-lib/ in order for act to use what is in that directory as the local repository?

This should not be a git based replacement, the tag is just what you specified in your yml.

And No the default old action cache is not supported together with --local-repository (layered on top of the new action cache!)

You need to remove the file path from the left side
Uses|CLI|Supported
---|---|---
`my-company/actions-lib/.github/workflows/build.yml@foo/initial`|`--local-repository my-company/actions-lib@foo/initial=<path>`|Yes
`my-company/actions-lib/.github/workflows/build.yml@foo/initial`|`--local-repository my-company/actions-lib/.github/workflows/build.yml@foo/initial=<path>`|No
`my-company/actions-lib@foo/initial`|`--local-repository my-company/actions-lib@foo/initial=<path>`|Yes

Shure there is room for improvement

**Comment by dusrdev** [comment]
@mverkerk-godaddy could you please explain why

> Error: stat /home/me/.cache/act/org-repo@main/.github/workflows/reusable.yml: no such file or directory

is expected? The error messages from act show that it should support remote reusable workflows, and all related issues/pr(s) have been merged.

At the time of writing (Act v0.2.79) I still get similar errors and thus enable to test workflows locally.

Even though all pr(s) related to running remote reusable workflows have been merged, none actually document how to actually use these "new features", and neither does the documentation or sample.

Has anyone found a way to be able to trigger a workflow locally that uses a remote reusable workflow? If so please let me know.
